pkg/errors/errors.go

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+//nolint
+// The idea of writing errors package in thanos is highly motivated from the Tast project of Chromium OS Authors. However, instead of
+// copying the package, we end up writing our own simplified logic borrowing some ideas from the errors and github.com/pkg/errors.
+// A big thanks to all of them.
+
+// Package errors provides basic utilities to manipulate errors with a useful stacktrace. It combines the
+// benefits of errors.New and fmt.Errorf world into a single package.
+package errors
+
+import (
+	//lint:ignore faillint Custom errors package needs to import standard library errors.
+	"errors"
+	"fmt"
+	"strings"
+)
+
+// base is the fundamental struct that implements the error interface and the acts as the backbone of this errors package.
+type base struct {
+	// info contains the error message passed through calls like errors.Wrap, errors.New.
+	info string
+	// stacktrace stores information about the program counters - i.e. where this error was generated.
+	stack stacktrace
+	// err is the actual error which is being wrapped with a stacktrace and message information.
+	err error
+}
+
+// Error implements the error interface.
+func (b *base) Error() string {
+	if b.err != nil {
+		return fmt.Sprintf("%s: %s", b.info, b.err.Error())
+	}
+	return b.info
+}
+
+// Unwrap implements the error Unwrap interface.
+func (b *base) Unwrap() error {
+	return b.err
+}
+
+// Format implements the fmt.Formatter interface to support the formatting of an error chain with "%+v" verb.
+// Whenever error is printed with %+v format verb, stacktrace info gets dumped to the output.
+func (b *base) Format(s fmt.State, verb rune) {
+	if verb == 'v' && s.Flag('+') {
+		s.Write([]byte(formatErrorChain(b)))
+		return
+	}
+
+	s.Write([]byte(b.Error()))
+}
+
+// New creates a new error with the given message and a stacktrace in details.
+// An alternative to errors.New function. It provides error message formatting
+// functionality i.e. a format string with different format verbs as input
+// arguments it acts similar to the fmt.Errorf (with a stacktrace).
+func New(format string, args ...interface{}) error {
+	return &base{
+		info:  fmt.Sprintf(format, args...),
+		stack: newStackTrace(),
+		err:   nil,
+	}
+}
+
+// Wrap creates a new error with the given message, wrapping another error and a stacktrace in details.
+// If cause is nil, this is the same as New. This Wrap method combines both Wrap and Wrapf functionality.
+// If wrapped is used with a format string with format verbs and the args, it will be same as Wrapf (a
+// convention used in error packages).
+func Wrap(cause error, format string, args ...interface{}) error {
```

Comment:
I wonder - there might be use cases where one wants to use stack traces always - thus doing something like `errors.Wrap(err)` always 🙃  - we could play with `Wrap(cause error, args ...interface{})` for this.

I think however this is the case is not worth optimization for - quite rare pattern and if we want you can always do another method for this.

But I think text is still useful - if you

Reply:
> But I think text is still useful 

Yes, absolutely. I am not sure if allowing the `errors.Wrap(err)` without a "reason string" would be a good idea or not. I think it would cause bad development practice as most of the time inside wrap the reason string would be omitted (just because it's optional :smile: ).